### DISCUSSION TITLE
Uruchamianie mikroserwisu do przetwarzania multimediów za pomocą usług Azure

### DISCUSSION THREAD

**Discussion by viters** [AzureCommunityPL/group#30] [Pomoc]
Hej!

Pracuję ostatnio nad mikroserwisem, który składa filmy ze zdjęć. W tym celu napisałem sobie aplikację w node.js, która działa mniej więcej tak:
- odbiera request HTTP z parametrami wideo i danymi
- odpowiada statusem 202
- pobiera zdjęcia
- składa film za pomocą ffmpeg
- wysyła film na Azure Storage
- wysyła callback HTTP do zlecającego

Obecnie uruchamiam go jako App Service na P1v2, prędkość wykonania zadania jest znośna i mogę skorzystać z automatycznego skalowania horyzontalnego. Natomiast mam wrażenie, że te App Service nie są idealne pod moje zastosowanie, gdyż w tym zadaniu przede wszystkim zależy mi na procesorze - nie potrzebuję dużo RAMu ani dużego dysku, a nie mogę wybrać VMki pod planem. Jednocześnie estymuję, że przy maksymalnie trzech instancjach i założeniu, że sumarycznie będą działać około 840 godzin miesięcznie (gdyż co najmniej 1 instancja musi działać przez cały miesiąc), zapłacę za to 81,47 € / mc. Wydaje mi się to dość dużą kwotą za prosty mikroserwis - dlatego szukam innych rozwiązań - niekoniecznie tańszych, ale być może zapewniających lepszą wydajność za te same pieniądze.

Testowałem Azure Functions, jednak w podstawowym planie nie ma wsparcia dla funkcji opartych o obrazy Dockera, a bazowe środowisko dla node.js nie ma zainstalowanego programu ffmpeg. Zastanawiałem się nad użyciem funkcji tylko do komunikacji, a samą konwersję zdjęć na wideo uruchomić w ramach innej usługi Azurowej, ale nie znam odpowiedniej. Kojarzę Azure Batch, ale wydaje mi się zbyt rozległa dla moich potrzeb. Widzę jeszcze opcję uruchomienia AKS i użycia węzłów zoptymalizowanych pod obliczenia (z serii F) - ale jest to najbardziej skomplikowana opcja pod kątem wdrożenia i w ramach kosztów trzeba też wziąć pod uwagę dodatkowe potrzebne usługi (jak np. certyfikaty SSL) oraz zarządzanie tym wdrożeniem.

Zakładam na razie niewielkie wykorzystanie tego mikroserwisu, do kilkudziesięciu requestów dziennie. Zastanawiam się, jaki jest najlepszy sposób (pod względem stosunku wydajności do ceny) uruchomienia takiego mikroserwisu na Azure.

Będę bardzo wdzięczny za komentarze / podpowiedzi.

**Comment by kamil-mrzyglod** [comment]
Patrzyłeś może na hhttps://docs.microsoft.com/en-us/azure/container-instances/ czy nie spełniają Twoich wymagań? AKS wygląda jak armata na muchę - nie prościej robić deployment po prostu VMki, zrobienie konwersji i jej ubicie?

**Reply by viters** [reply]
Masz rację, też wydaje mi się, że AKS to przesada. Nie trafiłem wcześniej na poleconą przez Ciebie usługę, na pewno się jej przyjrzę. Rozumiem, że flow byłby tam mniej więcej taki: zbudowałbym funkcję Azure, która odbiera request i uruchamia Azure Container Instance przekazując mu potrzebne dane, a resztę zadania wykonuje konkretna instancja, po czym się zamyka?

**Reply by kamil-mrzyglod** [reply]
Flow to już jak sobie zbudujesz, ale z grubsza tak, możesz spróbować uruchamiać kontener w ACI na podstawie zapytania, masz do tego nawet całe SDK dla NodeJS - https://docs.microsoft.com/en-us/javascript/api/overview/azure/containerinstances?view=azure-node-latest jeśli np. Twój Function App byłby hostowany wraz z tą platformą.

Masz nawet pokazane jak triggerować ACI z użyciem AF - https://docs.microsoft.com/en-us/azure/container-instances/container-instances-tutorial-azure-function-trigger

**Reply by wi5nia** [reply]
Tak w ogole co jeste staki uparty na Dockera w tym wszystkim? Tu same funkcje spokojnie to opedzluja ewentualnie z jakims wywolaniem czegos po srodku co obrobi te filmy.
Wiem, ze Docker i kontenery modne, ale nie ma co sie pchac z tym wszedzie.

**Reply by viters** [reply]
To nie musi być docker, nie mam ku temu żadnych wymagań. Właśnie tego czegoś pośrodku poszukuję, ale też zależy mi, żeby całość rozwiązania nie była bardzo skomplikowana i łatwo skalowalna.

**Reply by ptrstpp950** [reply]
Robiłem na wiosnę analizę do swojego PoC dla czegoś takiego (obrazki + tekst zamienione na video + dźwięk) i argumenty moje były takie (uwzględniając koszty, zarządzanie tym oraz możliwości):
- Azure Functions - odpadło wtedy ze względu na brak wsparcia do ffmpeg w planie consumption. Widziałem potem zgłoszony błąd i mieli to poprawić. Ważne jest też czy będziesz czy nie będziesz używał przejść OpenGL, bo one mają znaczący wpływ na "zużycie"
- App Plan - Drogo wychodziło (jak sam zauważyłeś) i dziwne mi się czasem ffmpeg+OpenGL zachowywał.
- ACI wspomniane przez @kamil-mrzyglod - całkiem spoko kosztowo, tylko wolno trwało. To znaczy czas na ściągnięcie obrazu docker z całym moim "stosem" narzędziowym był średni. Dla zastosowania wrzucę i puszczę spoko, dla zastosowania chce jak najszybciej już nie. Warto kilkanaście razy puścić sobie gotowy obraz do przetworzenia i zobaczyć czy wynik Cię satysfakcjonuje.
- VM - nie chce mi się użerać z VM'ką i updatem softu na VM'ce, więc ją odrzuciłem.
- AKS - de facto to samo co VM'ka tylko ładniej obudowane, więc lepsze niż czysta VM'ka. Jak ACI jest za wolne i Azure Functions mi nie naprawią to skończę tutaj. Nie ma też problemów z przejściami OpenGL i bardzo pomogło skalowanie puli do zera ;)

**Reply by wi5nia** [reply]
@ptrstpp950 a czy w ACI korzystales z maszyn ze wsparciem GPU czy bez?

**Reply by ptrstpp950** [reply]
@wi5nia Nie używałem GPU, bo dawało radę bez niego. Używało więcej CPU, ale to byłem wstanie przeżyć ze względu na różnicę w cenie 😉
p.s. odpowiadanie mailem na dyskusje na razie działa średnio bo nie dodaje do wątku :)

**Reply by wi5nia** [reply]
Odpowiadanie mailem na grupach w ogole slaba opcja ;)

No to sprobuj z GPU bo robi roznice jak Twoj soft czy cos czego uzywasz to obsluguje. No na ACI w zalozeniu na krtoka zytownosc powolywanych instancji wychodzi bardzo atrakcyjnie w takim scenariuszu.

**Reply by ptrstpp950** [reply]
Jak wrócę do projektu to spróbuje, acz głównym pożeraczem czasu było samo uruchomienie się ACI. Ponieważ odpalenie ffmpeg nawet z "magicznymi przejściami" dla 50-100 obrazków jest dość szybkie, a konfiguracja w Node.js wywołania ffmpeg tak aby używał GPU trywialna nie jest, to dałem sobie wtedy spokój :)

**Reply by wi5nia** [reply]
Nikt nie mowil, ze bedzie prosto ;) mozna jeszcze pomyslec o jakims Azure Batchu do tego typu zabaw no i to tez wszystko zalezy od tego jakie sa wzorce obciazenia, czy random, czy jednak przewidywalne, powtarzalne

**Reply by kmadof** [reply]
Co do ffmepga to ja mam właśnie na Azure Functions tyle że nie w Consumptions. I działa to dobrze. Jak do tej pory jedyne problemy jakie wychodziły to jak jakiś trefny plik trafił się do obróbki.
